pkg/querier/frontend/frontend.go

```diff
@@ -179,6 +179,16 @@ func (f *Frontend) Handler() http.Handler {
 }
 
 func (f *Frontend) handle(w http.ResponseWriter, r *http.Request) {
+	// to parse form we need to be sure
+	// that roundtriper request gets not read reader
+	var buf bytes.Buffer
+	r.Body = ioutil.NopCloser(io.TeeReader(r.Body, &buf))
+	// Ensure the form has been parsed so all the parameters are present
+	err := r.ParseForm()
+	if err != nil {
+		level.Warn(util.WithContext(r.Context(), f.log)).Log("msg", "unable to parse form for request", "err", err)
+	}
+	r.Body = ioutil.NopCloser(&buf)
 
```

Comment:
🤔  That'a a good question. In case of parsing FormData I don't know what we should do as it should be (❓ ) rare, a specially when we have eliminated case of reading closed body. So, to not break current implementation I would leave it as is and perhaps wait if someone reports something odd with this behaviour (which current implementation uses but in different order). And then think about solution, but if you think we should do something different I am happy to implement it.

In case of
```
if err != nil {
		writeError(w, err)
	}
``` 
I think we should still report slow query even in case of request error, we could also improve it by adding log entry the same as for slow query to indicate actual roundtrip error with additional context.

Reply:
> So, to not break current implementation I would leave it as is and perhaps wait if someone reports something odd with this behaviour

I don't think we would break the current implementation if we return in case of error, because we would just skipping logging slow queries, which doesn't make much sense if the query failed. Requests logging is done in a middleware so would be done anyway.

Reply:
This part of the code changed a bit, could you please see if it still makes sense ?